File: benchmarks/base64/c/base64.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static const char ENC_TABLE[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

static char *base64_encode(const unsigned char *data, size_t len, size_t *out_len) {
    size_t enc_len = ((len + 2) / 3) * 4;
    char *out = malloc(enc_len + 1);
    size_t i = 0, j = 0;
    while (i + 3 <= len) {
        unsigned char b0 = data[i], b1 = data[i + 1], b2 = data[i + 2];
        out[j++] = ENC_TABLE[b0 >> 2];
        out[j++] = ENC_TABLE[((b0 & 0x03) << 4) | (b1 >> 4)];
        out[j++] = ENC_TABLE[((b1 & 0x0F) << 2) | (b2 >> 6)];
        out[j++] = ENC_TABLE[b2 & 0x3F];
        i += 3;
    }
    size_t rem = len - i;
    if (rem == 1) {
        unsigned char b0 = data[i];
        out[j++] = ENC_TABLE[b0 >> 2];
        out[j++] = ENC_TABLE[(b0 & 0x03) << 4];
        out[j++] = '=';
        out[j++] = '=';
    } else if (rem == 2) {
        unsigned char b0 = data[i], b1 = data[i + 1];
        out[j++] = ENC_TABLE[b0 >> 2];
        out[j++] = ENC_TABLE[((b0 & 0x03) << 4) | (b1 >> 4)];
        out[j++] = ENC_TABLE[(b1 & 0x0F) << 2];
        out[j++] = '=';
    }
    out[j] = '\0';
    *out_len = j;
    return out;
}

static int dec_value(unsigned char c) {
    if (c >= 'A' && c <= 'Z') return c - 'A';
    if (c >= 'a' && c <= 'z') return c - 'a' + 26;
    if (c >= '0' && c <= '9') return c - '0' + 52;
    if (c == '+') return 62;
    if (c == '/') return 63;
    return -1;
}

static unsigned char *base64_decode(const char *enc, size_t enc_len, size_t *out_len) {
    unsigned char *out = malloc((enc_len / 4) * 3 + 3);
    size_t j = 0;
    for (size_t i = 0; i < enc_len; i += 4) {
        int v0 = dec_value((unsigned char)enc[i]);
        int v1 = dec_value((unsigned char)enc[i + 1]);
        int v2 = (enc[i + 2] == '=') ? -2 : dec_value((unsigned char)enc[i + 2]);
        int v3 = (enc[i + 3] == '=') ? -2 : dec_value((unsigned char)enc[i + 3]);
        out[j++] = (unsigned char)((v0 << 2) | (v1 >> 4));
        if (v2 != -2) {
            out[j++] = (unsigned char)(((v1 & 0x0F) << 4) | (v2 >> 2));
            if (v3 != -2) {
                out[j++] = (unsigned char)(((v2 & 0x03) << 6) | v3);
            }
        }
    }
    *out_len = j;
    return out;
}

int main(int argc, char **argv) {
    long n = argc > 1 ? atol(argv[1]) : 20000000;

    size_t elen;
    char *e;

    e = base64_encode((const unsigned char *)"", 0, &elen);
    if (elen != 0) { fprintf(stderr, "self-check failed: base64(\"\") mismatch\n"); return 1; }
    free(e);

    e = base64_encode((const unsigned char *)"f", 1, &elen);
    if (strcmp(e, "Zg==") != 0) { fprintf(stderr, "self-check failed: base64(\"f\") mismatch\n"); return 1; }
    free(e);

    e = base64_encode((const unsigned char *)"fo", 2, &elen);
    if (strcmp(e, "Zm8=") != 0) { fprintf(stderr, "self-check failed: base64(\"fo\") mismatch\n"); return 1; }
    free(e);

    e = base64_encode((const unsigned char *)"foo", 3, &elen);
    if (strcmp(e, "Zm9v") != 0) { fprintf(stderr, "self-check failed: base64(\"foo\") mismatch\n"); return 1; }
    free(e);

    unsigned char *buf = malloc((size_t)n);
    for (long i = 0; i < n; i++) buf[i] = (unsigned char)((i * 131 + 7) % 256);

    char *encoded = base64_encode(buf, (size_t)n, &elen);

    size_t dlen;
    unsigned char *decoded = base64_decode(encoded, elen, &dlen);
    if (dlen != (size_t)n || memcmp(decoded, buf, (size_t)n) != 0) {
        fprintf(stderr, "self-check failed: roundtrip mismatch\n");
        return 1;
    }

    long sum = 0;
    for (size_t i = 0; i < elen; i++) sum += (unsigned char)encoded[i];
    printf("%ld\n", sum);

    free(buf);
    free(encoded);
    free(decoded);
    return 0;
}

File: benchmarks/base64/rust/base64.rs
use std::env;

const ENC_TABLE: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

fn base64_encode(data: &[u8]) -> String {
    let mut out = Vec::with_capacity(((data.len() + 2) / 3) * 4);
    let mut i = 0;
    while i + 3 <= data.len() {
        let (b0, b1, b2) = (data[i], data[i + 1], data[i + 2]);
        out.push(ENC_TABLE[(b0 >> 2) as usize]);
        out.push(ENC_TABLE[(((b0 & 0x03) << 4) | (b1 >> 4)) as usize]);
        out.push(ENC_TABLE[(((b1 & 0x0F) << 2) | (b2 >> 6)) as usize]);
        out.push(ENC_TABLE[(b2 & 0x3F) as usize]);
        i += 3;
    }
    let rem = data.len() - i;
    if rem == 1 {
        let b0 = data[i];
        out.push(ENC_TABLE[(b0 >> 2) as usize]);
        out.push(ENC_TABLE[((b0 & 0x03) << 4) as usize]);
        out.push(b'=');
        out.push(b'=');
    } else if rem == 2 {
        let (b0, b1) = (data[i], data[i + 1]);
        out.push(ENC_TABLE[(b0 >> 2) as usize]);
        out.push(ENC_TABLE[(((b0 & 0x03) << 4) | (b1 >> 4)) as usize]);
        out.push(ENC_TABLE[((b1 & 0x0F) << 2) as usize]);
        out.push(b'=');
    }
    String::from_utf8(out).unwrap()
}

fn dec_value(c: u8) -> i32 {
    match c {
        b'A'..=b'Z' => (c - b'A') as i32,
        b'a'..=b'z' => (c - b'a') as i32 + 26,
        b'0'..=b'9' => (c - b'0') as i32 + 52,
        b'+' => 62,
        b'/' => 63,
        _ => -1,
    }
}

fn base64_decode(enc: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity((enc.len() / 4) * 3);
    let mut i = 0;
    while i < enc.len() {
        let v0 = dec_value(enc[i]);
        let v1 = dec_value(enc[i + 1]);
        let v2 = if enc[i + 2] == b'=' { -2 } else { dec_value(enc[i + 2]) };
        let v3 = if enc[i + 3] == b'=' { -2 } else { dec_value(enc[i + 3]) };
        out.push(((v0 << 2) | (v1 >> 4)) as u8);
        if v2 != -2 {
            out.push((((v1 & 0x0F) << 4) | (v2 >> 2)) as u8);
            if v3 != -2 {
                out.push((((v2 & 0x03) << 6) | v3) as u8);
            }
        }
        i += 4;
    }
    out
}

fn main() {
    let n: usize = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(20_000_000);

    assert_eq!(base64_encode(b""), "", "self-check failed: base64(\"\") mismatch");
    assert_eq!(base64_encode(b"f"), "Zg==", "self-check failed: base64(\"f\") mismatch");
    assert_eq!(base64_encode(b"fo"), "Zm8=", "self-check failed: base64(\"fo\") mismatch");
    assert_eq!(base64_encode(b"foo"), "Zm9v", "self-check failed: base64(\"foo\") mismatch");

    let mut buf = vec![0u8; n];
    for i in 0..n {
        buf[i] = ((i * 131 + 7) % 256) as u8;
    }

    let encoded = base64_encode(&buf);
    let decoded = base64_decode(encoded.as_bytes());
    assert_eq!(decoded, buf, "self-check failed: roundtrip mismatch");

    let sum: u64 = encoded.bytes().map(|b| b as u64).sum();
    println!("{}", sum);
}
